master/internal/resourcemanagers/priority.go

```diff
@@ -0,0 +1,331 @@
+package resourcemanagers
+
+import (
+	"fmt"
+	"sort"
+
+	log "github.com/sirupsen/logrus"
+
+	"github.com/determined-ai/determined/master/pkg/actor"
+	"github.com/determined-ai/determined/master/pkg/model"
+)
+
+type priorityScheduler struct {
+	preemptionEnabled bool
+}
+
+// NewPriorityScheduler creates a new scheduler that schedules tasks via priority.
+func NewPriorityScheduler(config *SchedulerConfig) Scheduler {
+	return &priorityScheduler{preemptionEnabled: config.Priority.Preemption}
+}
+
+func (p *priorityScheduler) Schedule(rp *ResourcePool) ([]*AllocateRequest, []*actor.Ref) {
+	return p.prioritySchedule(rp.taskList, rp.groups, rp.agents, rp.fittingMethod)
+}
+
+func (p *priorityScheduler) prioritySchedule(
+	taskList *taskList,
+	groups map[*actor.Ref]*group,
+	agents map[*actor.Ref]*agentState,
+	fittingMethod SoftConstraint,
+) ([]*AllocateRequest, []*actor.Ref) {
+	agentsSplitByLabel := splitAgentsByLabel(agents)
+	toAllocate := make([]*AllocateRequest, 0)
+	toRelease := make([]*actor.Ref, 0)
+
+	// Since labels are a hard scheduling constraint, process every
+	// label independently.
+	for label, agentsWithLabel := range agentsSplitByLabel {
+		toAllocatedForLabel, toReleaseForLabel := p.priorityScheduleByLabel(
+			taskList, groups, agentsWithLabel, fittingMethod, label)
+		toAllocate = append(toAllocate, toAllocatedForLabel...)
+		toRelease = append(toRelease, toReleaseForLabel...)
+	}
+
+	return toAllocate, toRelease
+}
+
+func (p *priorityScheduler) priorityScheduleByLabel(
+	taskList *taskList,
+	groups map[*actor.Ref]*group,
+	agents map[*actor.Ref]*agentState,
+	fittingMethod SoftConstraint,
+	label string,
+) ([]*AllocateRequest, []*actor.Ref) {
+	// All pending zero slot tasks get scheduled right away.
+	toAllocate := getAllPendingZeroSlotTasks(taskList, label)
```

Comment:
blocking: we should put preempting zero-slot tasks in the scope.

Reply:
I rebased and added this functionality.